Add container image sbom example generator

mprpic opened the pull request:
Also add an example SBOM for the ubi9-micro container image.

mprpic commented:
@twaugh Are there any other container images that would be good examples for this type that you'd recommend including here?

twaugh commented:
> @twaugh Are there any other container images that would be good examples for this type that you'd recommend including here?

It would definitely be good to have a simple example container that simply installs one or more of the example RPMs. That way we'll have complete example data to show _linking_ between them.

twaugh commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,222 @@
+        "creationInfo": {
+            "created": "2006-08-14T02:34:56-06:00",
+            "creators": [
+                "example SPDX document only",
```
```suggestion
                "Tool: example SPDX document only",
```

minor fix for strict checkers like TPA

twaugh commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,222 @@
+                # We don't have a way to find out which content set (RPM repo) an RPM came from,
+                # so we arbitrarily choose one here (assuming we have this mapping via RPM
+                # lockfiles or other means eventually).
+                f"arch={rpm['architecture']}&repository_url={content_sets[0]}"
```
Should this be `repository_id=`?

mprpic replied:
Ah yes! Fixed!

twaugh commented on sbom/examples/container_image/ubi9-micro-container-9.4-6.1716471860.spdx.json:
```diff
@@ -0,0 +1,2251 @@
+          "referenceCategory": "OTHER",
+          "referenceType": "sbom_ref",
+          "referenceLocator": "https://access.redhat.com/security/data/sbom/v1/rpm/gcc-0:11.4.1-3.el9.spdx.json.bz2"
+        }
```
I suppose this `sbom_ref` is a non-normative way to say "you can find the SBOM for that over here"?

The full external document linking style would be this: https://spdx.github.io/spdx-spec/v2.3/document-creation-information/#6.6
where we'd have things like `"relatedSpdxElement": "DocumentRef-rpm-gcc:SPDXRef-s390x-libgcc"` in `"relationships"` to describe relationships to packages not listed in this document.

twaugh replied:
And, to be clear, I think we don't necessarily want to use that full external document linking style, if purl matching is sufficient and provides more consumer convenience.

mprpic replied:
> I suppose this `sbom_ref` is a non-normative way to say "you can find the SBOM for that over here"?
> 

Yep!

> The full external document linking style would be this: https://spdx.github.io/spdx-spec/v2.3/document-creation-information/#6.6 where we'd have things like `"relatedSpdxElement": "DocumentRef-rpm-gcc:SPDXRef-s390x-libgcc"` in `"relationships"` to describe relationships to packages not listed in this document.

Oh neat, I didn't realize this existed, but it does create a lot more objects in the SBOM since you need a separate `DocumentRef` and relationship object for each RPM.

The CycloneDX approach also seems to be similar to the one I used here where there is a reference type called `bom` to point to other SBOMs: https://cyclonedx.org/use-cases/#external-references

Do you have a preference for one or the other?

> And, to be clear, I think we don't necessarily want to use that full external document linking style, if purl matching is sufficient and provides more consumer convenience.

purl matching works when you import it into something like Guac/Trustification, but if we are to also provide these documents as is for anyone to download and parse on their own, we'd be asking them to also re-implement the matching of purls themselves to discover the links between different artifacts. Providing the reference(s) in the document would make that a lot easier, I'd think.

Ideally, our VEX and advisory files would also link / refer to the SBOMs so that we could potentially reduce their size to just be one or more references to actual SBOM files instead of replicating the information between the two.

twaugh replied:
> Providing the reference(s) in the document would make that a lot easier, I'd think.

+1

mprpic replied:
I removed the `sbom_ref` reference since we'll be relying on purl matching only.

mprpic commented:
> It would definitely be good to have a simple example container that simply installs one or more of the example RPMs. That way we'll have complete example data to show _linking_ between them.
> 

I added an example for the `kmm/kernel-module-management-rhel9-operator` that installs the same version of openssl as we have in our rpm example. It does also build and install a Go binary but I'm not sure where to fetch that data from. Any ideas?

I also added an example of the podman container, which just contains extra RPMs on top of the ones installed in UBI. Though, we don't really make a distinction between the RPMs installed in the base layer vs the final layer (or any layers inbetween for that matter).

twaugh commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,229 @@
+
+        for rpm in get_rpms(catalog_image_id):
+            purl = (
+                f"pkg:rpm/redhat/{rpm['name']}@{rpm['version']}?"
```
```suggestion
                f"pkg:rpm/redhat/{rpm['name']}@{rpm['version']}-{rpm['release']}?"
```

twaugh commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,229 @@
+            repo_name = repo["repository"]
+            repo_namespace, _, repo_name = repo_name.rpartition("/")
+            if repo_namespace:
+                registry = f"{registry}/{repo_namespace}"
```
I think the repository_url is not intended to be the namespace but the full registry/namespace/name part. See the examples here: https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst

So for example, I think we want output like this:
```
"externalRefs": [
  {
    "referenceCategory": "PACKAGE-MANAGER",
    "referenceType": "purl",
    "referenceLocator": "pkg:oci/podman@sha256%3A749dc1a252dfec536ae15bbb4f509a3bccb40de6abf9712619cb8df2d99e94b7?repository_url=registry.access.redhat.com/rhel9/podman&tag=9.4-8"
  },
```

mprpic replied:
Mmm, you're right, nice catch! I also need to fix that in the purl guidelines doc.

mprpic replied:
https://github.com/RedHatProductSecurity/security-data-guidelines/pull/7

twaugh commented on sbom/examples/container_image/ubi9-micro-container-9.4-6.1716471860.spdx.json:
```diff
@@ -21,12 +21,12 @@
           "referenceCategory": "PACKAGE-MANAGER",
           "referenceType": "purl",
-          "referenceLocator": "pkg:oci/ubi9-micro@sha256%3A1c8483e0fda0e990175eb9855a5f15e0910d2038dd397d9e2b357630f0321e6d?repository_url=registry.access.redhat.com&tag=9.4-6.1716471860"
+          "referenceLocator": "pkg:oci/ubi9-micro@sha256%3A1c8483e0fda0e990175eb9855a5f15e0910d2038dd397d9e2b357630f0321e6d?repository_url=registry.access.redhat.com//ubi9-micro&tag=9.4-6.1716471860"
```
I wonder why this has "//"?

mprpic replied:
Ah, that's because ubi9-micro doesn't have a namespace so it was empty. Fixed it!

mprpic commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,232 @@
+                    {
+                        "referenceCategory": "PACKAGE-MANAGER",
+                        "referenceType": "purl",
+                        "referenceLocator": srpm_purl,
```
I added a purl for the SRPM that this binary RPM belongs to, so that the link exists between both the binary component and the SRPM component, but perhaps it's not needed? Do we even have information about SRPMs during the build of a container image that contains just binary RPMs?

twaugh replied:
I think we don't, and I think it's not needed.

mprpic commented on sbom/examples/container_image/from_catalog.py:
```diff
@@ -0,0 +1,232 @@
+            if label["name"].lower() == "license":
+                image_license = label["value"]
+
+        # TODO: split image index into its own SBOM and produce per-arch image SBOM
```
I'll rework the logic below to produce one SBOM file for each architecture with the root node being the arch-specific image, and one SBOM file that describes the image list and its descendant arch-specific images.

mprpic replied:
@twaugh Alright, this is done. We now have one SBOM that represents the image index, and four separate SBOMs, one for each arch. Can you please give this one last look before I merge it?

In a separate PR I can create a separate example that highlights referred vs embedded content.

mprpic replied:
Going to merge this as is. We can further refine the container examples in separate PRs if needed.
